Automattic/woocommerce-payments
# Express button product page

#### Changes proposed in this Pull Request
This PR implements WooPay express checkout button on the Product page and it works the same as Google Pay/Apple Pay. On WooPay express button click it adds that product to the merchant's cart first and then initiates the platform checkout. Different quantities of the product, different variation and add-ons also works exactly same as other payment request buttons.

Some notable changes are-
- Added a hook `use-express-checkout-product-handler.js` which handles the product page related extra pre-processing. The logics and functions here are heavily borrowed from [payment-request file](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-request/index.js#L134). The only difference is we are not using jquery in the hook. Otherwise, the logics depend on the status of the same dom elements as the payment request button.
- had to make a change in the `checkout/utils/request.js` to properly translate a nested object to form data.

#### Testing instructions
- Install the latest [dev-tools](https://github.com/Automattic/woocommerce-payments-dev-tools). Go to the dev tools page and click the Enable the WooPay Express Checkout button checkbox and save the settings.
- Go to the WooCommerce Payments settings screen and click the Customize button of WooPay.
- Select `Product` page as the button location and save the settings.
- Go to a simple product page and ensure that you can checkout with WooPay express checkout button.
- Go to a variable product page, select different variations and ensure that you can checkout with WooPay express checkout button.
- Clicking the WooPay express checkout button adds that product to the merchant cart and WooPay checkout should have the same product in the cart (with the selected quantity)
- Install the `Product Add On` extension, add some add-ons to a product. Go to that product page, select some add-ons and click WooPay express checkout button. Ensure that the cart has this product with the add-ons.
